uat/custom-components/client-python-paho/src/grpc_client_server/grpc_discovery_client.py

```diff
@@ -0,0 +1,148 @@
+#
+# Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+# SPDX-License-Identifier: Apache-2.0
+#
+
+"""Client of gRPC MqttAgentDiscovery service."""
+# pylint: disable=no-member,no-name-in-module
+import logging
+
+import grpc
+from exceptions.grpc_exception import GRPCException
+from grpc_client_server.grpc_generated.mqtt_client_control_pb2 import Mqtt5Disconnect, MqttConnectionId, Mqtt5Message
+from grpc_client_server.grpc_generated import mqtt_client_control_pb2_grpc, mqtt_client_control_pb2
+
+
+GRPC_REQUEST_FAILED = "gRPC request failed"
+
+
+class GRPCDiscoveryClient:
+    """Client of gRPC MqttAgentDiscovery service."""
+
+    logger = logging.getLogger("GRPCDiscoveryClient")
+
+    def __init__(self, agent_id: str, address: str):
+        """
+        Construct GRPCDiscoveryClient
+        Parameters
+        ----------
+        agentId - id of agent to identify control channel by gRPC server
+        address - address of gRPC service, including port
+        """
+        self.__logger = GRPCDiscoveryClient.logger
+        self.__agent_id = agent_id
+        self.__channel = grpc.insecure_channel(address)

```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**
This line of code assigns a resource that might leak to a class field, causing a resource leak that slows down or crashes your system. For example, when *mqtt\_client\_control\_pb2\_grpc.MqttAgentDiscoveryStub()* throws an exception, the *GRPCDiscoveryClient* object isn’t constructed so you can’t close this resource afterward.

**Fix**
Ensure that if an exception is thrown, the *\_\_channel* resource is closed before the constructor is exited.